quiche/src/lib.rs

```diff
@@ -1710,13 +1758,39 @@ impl Connection {
 
         let active_path_id = paths.get_active_path_id()?;
 
-        let ids = cid::ConnectionIdentifiers::new(
+        let mut ids = cid::ConnectionIdentifiers::new(
             config.local_transport_params.active_conn_id_limit as usize,
             scid,
             active_path_id,
             reset_token,
         );
 
+        // If we aren't using zero length connection id
+        if !scid.is_empty() {
+            // Track the scid if we sent one in the preferred address transport
+            // params.
+            if let Some(preferred_address_params) =
+                &config.local_transport_params.preferred_address_params
+            {
+                let seq = ids.new_scid(
+                    preferred_address_params.connection_id.to_vec().into(),
+                    Some(preferred_address_params.stateless_reset_token),
+                    // Don't advertise this because it's in the transport params.
+                    false,
+                    None,
+                    false,
+                )?;
+
+                // This should be sequence number one by RFC
+                debug_assert_eq!(seq, 1);
+            }
+        }
+
+        // Only servers support sending a preferred address in the transport paramters
+        if !is_server && config.local_transport_params.preferred_address_params.is_some() {
+            return Err(Error::InvalidTransportParam)
+        }
```

Comment:
We can do this check inside the `if let Some block` above. Then it might be a good candidate to extract into its own function anyway.